File: src/agentJson.ts
import { Buffer } from "node:buffer";

export const AGENT_JSON_SCHEMA_VERSION = 1 as const;
export const AGENT_JSON_MAX_BYTES = 64 * 1024;

export const SUPERCOV_ERROR_CODES = [
  "AMBIGUOUS_SELECTOR",
  "DECISION_NOT_FOUND",
  "FILTER_UNAVAILABLE",
  "INTERNAL_ERROR",
  "INVALID_ARGUMENT",
  "MINIMIZATION_COMPLEXITY_LIMIT",
  "NO_RUNS",
  "RESPONSE_TOO_LARGE",
  "RUN_NOT_FOUND",
  "SCOPE_UNAVAILABLE",
  "SOURCE_NOT_FOUND",
  "TARGET_UNREACHABLE",
  "TEST_FILTER_EMPTY",
  "TEST_NOT_FOUND",
  "UNATTRIBUTED_EVIDENCE",
  "UNKNOWN_COMMAND",
] as const;

export type SupercovErrorCode = (typeof SUPERCOV_ERROR_CODES)[number];

export interface AgentJsonPagination {
  offset: number;
  limit: number;
  returned: number;
  total: number;
  hasMore: boolean;
  nextOffset: number | null;
}

export interface AgentJsonSuccess {
  schemaVersion: typeof AGENT_JSON_SCHEMA_VERSION;
  ok: true;
  command: string;
  data: unknown;
  pagination?: AgentJsonPagination;
}

export interface AgentJsonFailure {
  schemaVersion: typeof AGENT_JSON_SCHEMA_VERSION;
  ok: false;
  command?: string;
  error: {
    code: SupercovErrorCode;
    message: string;
    retryable: boolean;
    details?: Record<string, unknown>;
  };
}

export class SupercovError extends Error {
  readonly code: SupercovErrorCode;
  readonly retryable: boolean;
  readonly details?: Record<string, unknown>;

  constructor(
    code: SupercovErrorCode,
    message: string,
    options: {
      retryable?: boolean;
      details?: Record<string, unknown>;
      cause?: unknown;
    } = {},
  ) {
    super(message, options.cause === undefined ? undefined : { cause: options.cause });
    this.name = "SupercovError";
    this.code = code;
    this.retryable = options.retryable ?? false;
    this.details = options.details;
  }
}

export function agentPagination(
  offset: number,
  limit: number,
  returned: number,
  total: number,
): AgentJsonPagination {
  const nextOffset = offset + returned;
  const hasMore = returned > 0 && nextOffset < total;
  return {
    offset,
    limit,
    returned,
    total,
    hasMore,
    nextOffset: hasMore ? nextOffset : null,
  };
}

function serialized(value: unknown): string {
  return `${JSON.stringify(value)}\n`;
}

export function agentSuccessJson(
  command: string,
  data: unknown,
  pagination?: AgentJsonPagination,
): string {
  const envelope: AgentJsonSuccess = {
    schemaVersion: AGENT_JSON_SCHEMA_VERSION,
    ok: true,
    command,
    data,
    ...(pagination ? { pagination } : {}),
  };
  const json = serialized(envelope);
  const bytes = Buffer.byteLength(json);
  if (bytes > AGENT_JSON_MAX_BYTES) {
    throw new SupercovError(
      "RESPONSE_TOO_LARGE",
      `JSON response is ${bytes} bytes; the maximum is ${AGENT_JSON_MAX_BYTES} bytes`,
      {
        details: {
          actualBytes: bytes,
          maxBytes: AGENT_JSON_MAX_BYTES,
          hint: "Use --offset/--limit or a narrower coverage query.",
        },
      },
    );
  }
  return json;
}

export function asSupercovError(error: unknown): SupercovError {
  if (error instanceof SupercovError) return error;
  return new SupercovError(
    "INTERNAL_ERROR",
    error instanceof Error ? error.message : String(error),
    { cause: error },
  );
}

export function agentFailureJson(
  error: unknown,
  command?: string,
): string {
  const failure = asSupercovError(error);
  const envelope: AgentJsonFailure = {
    schemaVersion: AGENT_JSON_SCHEMA_VERSION,
    ok: false,
    ...(command ? { command } : {}),
    error: {
      code: failure.code,
      message: failure.message,
      retryable: failure.retryable,
      ...(failure.details ? { details: failure.details } : {}),
    },
  };
  const json = serialized(envelope);
  if (Buffer.byteLength(json) <= AGENT_JSON_MAX_BYTES) return json;
  return serialized({
    schemaVersion: AGENT_JSON_SCHEMA_VERSION,
    ok: false,
    ...(command ? { command } : {}),
    error: {
      code: failure.code,
      message: failure.message.slice(0, 1_000),
      retryable: failure.retryable,
    },
  } satisfies AgentJsonFailure);
}

File: crates/supercov-engine/src/agent_json.rs
use serde::Serialize;
use serde_json::Value;
use supercov_contracts::{AGENT_JSON_MAX_BYTES, AGENT_JSON_SCHEMA_VERSION, AgentPagination};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
pub enum ErrorCode {
    AmbiguousSelector,
    DecisionNotFound,
    FilterUnavailable,
    InternalError,
    InvalidArgument,
    MinimizationComplexityLimit,
    NoRuns,
    ResponseTooLarge,
    RunNotFound,
    ScopeUnavailable,
    SourceNotFound,
    TargetUnreachable,
    TestFilterEmpty,
    TestNotFound,
    UnattributedEvidence,
    UnknownCommand,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentError {
    pub code: ErrorCode,
    pub message: String,
    pub retryable: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub details: Option<Value>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ResponseTooLarge {
    pub actual_bytes: usize,
    pub max_bytes: usize,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct SuccessEnvelope<'a, T> {
    schema_version: u32,
    ok: bool,
    command: &'a str,
    data: &'a T,
    #[serde(skip_serializing_if = "Option::is_none")]
    pagination: Option<&'a AgentPagination>,
}

#[derive(Serialize)]
#[serde(rename_all = "camelCase")]
struct FailureEnvelope<'a> {
    schema_version: u32,
    ok: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    command: Option<&'a str>,
    error: &'a AgentError,
}

fn newline_json<T: Serialize>(value: &T) -> Result<String, serde_json::Error> {
    let mut json = serde_json::to_string(value)?;
    json.push('\n');
    Ok(json)
}

pub fn pagination(offset: usize, limit: usize, returned: usize, total: usize) -> AgentPagination {
    let next = offset.saturating_add(returned);
    let has_more = returned > 0 && next < total;
    AgentPagination {
        offset,
        limit,
        returned,
        total,
        has_more,
        next_offset: has_more.then_some(next),
    }
}

pub fn success<T: Serialize>(
    command: &str,
    data: &T,
    page: Option<&AgentPagination>,
) -> Result<String, ResponseTooLarge> {
    let json = newline_json(&SuccessEnvelope {
        schema_version: AGENT_JSON_SCHEMA_VERSION,
        ok: true,
        command,
        data,
        pagination: page,
    })
    .expect("serializing a Supercov JSON envelope must not fail");
    if json.len() > AGENT_JSON_MAX_BYTES {
        return Err(ResponseTooLarge {
            actual_bytes: json.len(),
            max_bytes: AGENT_JSON_MAX_BYTES,
        });
    }
    Ok(json)
}

pub fn failure(command: Option<&str>, error: &AgentError) -> String {
    let envelope = FailureEnvelope {
        schema_version: AGENT_JSON_SCHEMA_VERSION,
        ok: false,
        command,
        error,
    };
    let json = newline_json(&envelope).expect("serializing a Supercov error must not fail");
    if json.len() <= AGENT_JSON_MAX_BYTES {
        return json;
    }

    let mut truncated = error.clone();
    truncated.message = truncated.message.chars().take(1_000).collect();
    truncated.details = None;
    newline_json(&FailureEnvelope {
        error: &truncated,
        ..envelope
    })
    .expect("serializing a truncated Supercov error must not fail")
}

#[cfg(test)]
mod tests {
    use serde_json::json;

    use super::*;

    #[test]
    fn success_is_byte_identical_to_the_reference_engine() {
        let actual = success(
            "coverage.summary",
            &json!({"run": "run-123", "coverage": {"lines": 100}}),
            None,
        )
        .unwrap();
        assert_eq!(
            actual,
            include_str!("../../../tests/golden/agent-success.json")
        );
    }

    #[test]
    fn pagination_is_byte_identical_to_the_reference_engine() {
        let page = pagination(20, 20, 1, 21);
        let actual = success(
            "coverage.gaps",
            &json!({"gaps": [{"file": "src/example.ts"}]}),
            Some(&page),
        )
        .unwrap();
        assert_eq!(
            actual,
            include_str!("../../../tests/golden/agent-page.json")
        );
    }

    #[test]
    fn errors_are_byte_identical_to_the_reference_engine() {
        let actual = failure(
            Some("coverage.file"),
            &AgentError {
                code: ErrorCode::SourceNotFound,
                message: "Source file not found: missing.ts".into(),
                retryable: false,
                details: Some(json!({"selector": "missing.ts"})),
            },
        );
        assert_eq!(
            actual,
            include_str!("../../../tests/golden/agent-error.json")
        );
    }

    #[test]
    fn success_enforces_the_agent_context_budget() {
        let result = success("coverage.file", &"x".repeat(AGENT_JSON_MAX_BYTES), None);
        assert!(matches!(
            result,
            Err(ResponseTooLarge {
                max_bytes: AGENT_JSON_MAX_BYTES,
                ..
            })
        ));
    }
}
